File: algorithms/M/Matrix Cells in Distance Order/Matrix Cells in Distance Order.py
class Solution:
    def allCellsDistOrder(self, rows: int, cols: int, rCenter: int, cCenter: int) -> List[List[int]]:
        # create a r, c matrix given the rows & cols
        # each element represents a list [r, c] where r is the row & c the col
        # find find the distances of all cells from the center (append to res)
        # sort the result by distance function
        # Time O(M + N) Space O(M + N)
        
        
        def distance(p1, p2):
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
        
        matrix = [[i, j] for i in range(rows) for j in range(cols)]
        center = [rCenter, cCenter]
        matrix.sort(key=lambda c: distance(c, center))
        
        return matrix

File: scripts/algorithms/M/Matrix Cells in Distance Order/Matrix Cells in Distance Order.cpp
class Solution {
public:
    vector<vector<int>> allCellsDistOrder(int rows, int cols, int rCenter, int cCenter) {
        vector<vector<int>> ans;
        multimap<int,vector<int>> mp;
        for(int i = 0;i<rows;i++){
            for(int j = 0;j<cols;j++){
                    mp.insert(pair<int,vector<int>>(abs(i-rCenter)+abs(j-cCenter),{i,j})); //map will sort the value based on key(distance)    
                }
            }
        for(auto itr = mp.begin();itr != mp.end();itr++){
            ans.push_back(itr->second);
        }
        return ans;
    }
};
